flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1405,7 +1440,31 @@ class ScalarExprLowering {
             });
         caller.placeInput(arg, boxChar);
       } else if (arg.passBy == PassBy::Box) {
-        caller.placeInput(arg, builder.createBox(getLoc(), argRef));
+        auto box = builder.createBox(loc, argRef);
+        // Before lowering to an address, handle the allocatable/pointer actual
+        // argument to optional fir.box dummy. It is legal to pass
+        // unallocated/disassociated entity to an optional. In this case, an
+        // absent fir.box must be created instead of a fir.box with a null value
+        // (Fortran 2018 15.5.2.12 point 1).
+        const auto *sym = Fortran::evaluate::UnwrapWholeSymbolDataRef(*expr);
+        if (arg.isOptional && sym &&
+            Fortran::semantics::IsAllocatableOrPointer(*sym)) {
+          // FIXME: the case where sym is itself an optional argument that is
+          // not present should also be taken into account. A additional
+          // selectOp is not enough to implement this because the allocatable
+          // descriptor should not be read in all cases. Control flow must be
+          // added to cover this case.
```

Comment:
Shouldn't `genMutableBoxValue` handle that case as it can just occur in any expression?

That is, any allocatble/pointer symbol might be in a disassociated state where its value is `absent`. This may imply we need to deal with a boxed `absent` or selectively box only non-`absent` values, I suppose.

Reply:
No, there is a pile of restriction around absent optional (F2018 15.5.2.12 point 3). Basically it is forbidden to do anything with absent optional, except passing it to PRESENT and to an optional dummy. And it is up to the programmer to deal with it (I am not quite sure what we could do anyway, except adding some runtime checks to gracefully crash).

And, reading at it more in detail, I realize my TODO is useless :) 
"(8) If it is allocatable, it shall  not [...] be supplied as an actual argument corresponding to an optional nonallocatable dummy argument".

So it is left to the programmer to not call a function with an absent allocatable/pointer passed to an optional non allocatable/pointer.

I have replaced the TODO with the standard reference explaining that we can unconditionally read the mutable box here, since it was not obvious to me at first.